### PR TITLE
feat: Enable inline PyPDF file_processors provider

### PR DESCRIPTION
**Pull request by alinaryan**
# What does this PR do?
This PR wires up the PyPDF file processor provider so it is registered with the starter distribution. With this change, the starter distro exposes the file processor route and successfully processes PDFs through the inline::pypdf provider.

Key behaviors verified:

starter distro includes file_processors API + inline::pypdf provider in the stack config

route /v1alpha/file-processors/process is present and backed by inline::pypdf

endpoint is described in openapi.json

PDFs can be processed both by direct upload to the file_processors API (file=@...pdf) and via uploaded file storage (file_id=...) to the files API.

## Test Plan

1. In a venv, run the starter distro:
```
export OLLAMA_URL="http://localhost:11434/v1"
llama stack run starter
```

2. Verify route is registered

```
curl -sS http://localhost:8321/v1/inspect/routes \
  | jq -r '.data[] | select(.route|test("file-processors";"i"))'
```

Expected output:

```
{
  "route": "/v1alpha/file-processors/process",
  "method": "POST",
  "provider_types": [
    "inline::pypdf"
  ]
}
```

3. Verify OpenAPI contains the endpoint + schema
`curl -sS http://localhost:8321/openapi.json | rg -n "/v1alpha/file-processors/process|file-processors" -i
`

4. Exercise endpoint via direct file upload (PDF)
```
curl -sS -X POST http://localhost:8321/v1alpha/file-processors/process \
  -F "file=@/pdfs//US-Youth-Soccer-Travel-Policy.pdf" \
  | jq
```

Output (excerpt):

```
{
  "chunks": [
    {
      "chunk_id": "US-Youth-Soccer-Travel-Policy.pdf_chunk_0",
      "metadata": {
        "filename": "US-Youth-Soccer-Travel-Policy.pdf",
        "page_count": 8
      }
    }
  ],
  "metadata": {
    "processor": "pypdf",
    "page_count": 8,
    "extraction_method": "pypdf"
  }
}
```

5. Exercise endpoint via file storage upload + file_id

Upload:
```

FILE_ID="$(
  curl -sS -X POST http://localhost:8321/v1/files \
    -F "file=@/path/to/pdf.pdf" \
    -F "purpose=assistants" \
  | jq -r .id
)"
echo "$FILE_ID"
```

Process via file_id:

```
curl -sS -X POST http://localhost:8321/v1alpha/file-processors/process \
  -F "file_id=${FILE_ID}" \
  | jq
```

Expected: similar successful chunks response with metadata including the file_id, and top-level metadata.processor = "pypdf".

**Notes / Known behaviors observed during testing**

Posting application/json with {"file_id": ...} returned "Either file or file_id must be provided" because the endpoint expects multipart/form-data for this operation (per OpenAPI). Verified correct behavior using -F file_id=....

Attempting to process a non-PDF (just-text.txt) returned an internal server error in this environment.

### CODE REVIEW

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        file_id: str | None = None,
+        options: dict[str, Any] | None = None,
+        chunking_strategy: VectorStoreChunkingStrategy | None = None,
+    ) -> ProcessFileResponse:
+        """Process a PDF file and return chunks."""
+
+        # Note: options parameter is currently unused for PyPDF processor
+        # All processing options are configured via PyPDFFileProcessorConfig
+        if options:
+            log.debug("PyPDFFileProcessor ignoring options=%s (MVP)", options)
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: is this necessary?

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+            # Read from uploaded file
+            content = await file.read()
+            filename = file.filename or "uploaded_file.pdf"
+        elif file_id:
+            # Get file from file storage using Files API
+            if not self.files_api:
+                raise ValueError("Files API not available - cannot process file_id")
+
+            # Import here to avoid circular dependencies
+            from llama_stack_api.files import RetrieveFileContentRequest, RetrieveFileRequest
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: are you sure this is a circular import? I don't think it is.

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 1 above)
Comment (alinaryan) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: no, removing

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 2 above)
Comment (alinaryan) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: removing that comment

File: docs/docs/providers/inference/remote_openai.mdx [hunk 3]
```diff
@@ -34,11 +34,11 @@ OpenAI inference provider for accessing GPT models and other OpenAI services.
 | `network.timeout.connect` | `float \| None` | No |  | Connection timeout in seconds. |
 | `network.timeout.read` | `float \| None` | No |  | Read timeout in seconds. |
 | `network.headers` | `dict[str, str] \| None` | No |  | Additional HTTP headers to include in all requests. |
-| `base_url` | `HttpUrl \| None` | No | https://api.openai.com/v1 | Base URL for OpenAI API |
+| `base_url` | `HttpUrl \| None` | No | https://api.openai.com/ | Base URL for OpenAI API |
```
Comment (cdoern) [inline docs/docs/providers/inference/remote_openai.mdx]: this is necessary, we used to append it for the user but no longer do.

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        if reader.is_encrypted:
+            if self.config.password:
+                reader.decrypt(self.config.password)
+            else:
+                raise ValueError("PDF is encrypted but no password provided")
+
+        metadata: dict[str, Any] = {"page_count": len(reader.pages)}
+
+        # Extract document metadata
+        if reader.metadata:
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:152]:
```suggestion
        if reader.metadata:
            pdf_metadata = reader.metadata
        
            fields = {
                "title": pdf_metadata.title,
                "author": pdf_metadata.author,
                "subject": pdf_metadata.subject,
                "creator": pdf_metadata.creator,
                "producer": pdf_metadata.producer,
                "creation_date": pdf_metadata.creation_date,
                "modification_date": pdf_metadata.modification_date,
            }
        
            metadata.update(
                {key: str(value) for key, value in fields.items() if value}
            )        
```

File: tests/integration/file_processors/test_pypdf_processor.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+
+    async def test_file_id_without_files_api(self, processor: PyPDFFileProcessor):
+        """Test processing file_id without files API."""
+        with pytest.raises(ValueError, match="Files API not available"):
+            await processor.process_file(file_id="test_file_id")
+
+    def _create_minimal_pdf(self) -> bytes:
+        """Create a minimal valid PDF for testing edge cases."""
+        # Minimal PDF content
+        pdf_content = b"""%PDF-1.4
```
Comment (franciscojavierarceo) [inline tests/integration/file_processors/test_pypdf_processor.py]: let's move this to a separate file and import it, yeah?

Comment (franciscojavierarceo) [reply tests/integration/file_processors/test_pypdf_processor.py]: or just move it to the top as a constnat

File: tests/integration/file_processors/test_pypdf_processor.py [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+        # Should raise an exception for invalid PDF
+        with pytest.raises(
+            (ValueError, RuntimeError, OSError, Exception)
+        ):  # PyPDF will raise various exceptions for invalid content including PdfStreamError
+            await processor.process_file(file=upload_file, chunking_strategy=None)
+
+    async def test_empty_pdf_handling(self, processor: PyPDFFileProcessor):
+        """Test handling of empty or contentless PDF."""
+        # Create a PDF with no text content
+        empty_pdf_content = b"""%PDF-1.4
```
Comment (franciscojavierarceo) [inline tests/integration/file_processors/test_pypdf_processor.py]: same here about constant

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 4 above)
Comment (franciscojavierarceo) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:152]: alternative (maybe cleaner)
```suggestion
        if reader.metadata:
            pdf_metadata = reader.metadata
        
            keys = [
                "title",
                "author",
                "subject",
                "creator",
                "producer",
                "creation_date",
                "modification_date",
            ]
        
            for key in keys:
                value = getattr(pdf_metadata, key, None)
                if value:
                    metadata[key] = str(value)
```

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+        for line in lines:
+            # Strip whitespace and normalize
+            cleaned_line = " ".join(line.split())
+            if cleaned_line:  # Only keep non-empty lines
+                cleaned_lines.append(cleaned_line)
+
+        return "\n".join(cleaned_lines)
+
+    def _create_chunks(
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:186]: is this porting over the old code?

File: src/llama_stack/providers/remote/inference/openai/config.py [hunk 8]
```diff
@@ -22,15 +22,15 @@ class OpenAIProviderDataValidator(BaseModel):
...
+        default=HttpUrl("https://api.openai.com"),
         description="Base URL for OpenAI API",
     )
 
     @classmethod
     def sample_run_config(
         cls,
         api_key: str = "${env.OPENAI_API_KEY:=}",
-        base_url: str = "${env.OPENAI_BASE_URL:=https://api.openai.com/v1}",
+        base_url: str = "${env.OPENAI_BASE_URL:=https://api.openai.com}",
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/remote/inference/openai/config.py]: as charlie noted, please add this back

**Comment by cdoern** [comment]
I think all of these failures are from being out of sync. pre-commit is trying to make a bunch of changes not related to your PR. you might have done an improper rebase or just are behind main. please rebase and run pre-commit locally before pushing to the PR.

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 7 above)
Comment (alinaryan) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:186]: Just refactored this to stop manually constructing chunks and now reuse `make_overlapped_chunks()` for both the single-chunk and overlapped cases and so that this is consistent with the provider design.

File: src/llama_stack/providers/registry/file_processors.py [hunk 9]
```diff
@@ -4,8 +4,40 @@
...
+    return [
+        InlineProviderSpec(
+            api=Api.file_processors,
+            provider_type="inline::pypdf",
+            pip_packages=["pypdf"],
+            module="llama_stack.providers.inline.file_processor.pypdf",
+            config_class="llama_stack.providers.inline.file_processor.pypdf.PyPDFFileProcessorConfig",
+            optional_api_dependencies=[Api.files],
+            description="""
+PyPDF file processor for extracting text content from PDF documents.
```
Comment (cdoern) [inline src/llama_stack/providers/registry/file_processors.py]: this is kind of lengthy, I might be wrong but most provider keep this short so the output looks nice

File: tests/integration/responses/recordings/175a594469fd2c66751bd037a78fa3d6453e7b8bbb2a55f7baf6147997026931.json
Comment (cdoern) [inline tests/integration/responses/recordings/175a594469fd2c66751bd037a78fa3d6453e7b8bbb2a55f7baf6147997026931.json:1]: these recordings are because you changed some openai utils, right? we should double check that

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        if file and file_id:
+            raise ValueError("Cannot provide both file and file_id")
+
+        start_time = time.time()
+
+        # Get PDF content
+        if file:
+            # Read from uploaded file
+            content = await file.read()
+            filename = file.filename or "uploaded_file.pdf"
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: probably we should generate a unique filename.

File: src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py
Comment (franciscojavierarceo) [inline src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py:107]:
```suggestion
        vector_stores_config: VectorStoresConfig | None = None,
        file_processor_api: FileProcessor | None = None,
```

Comment (franciscojavierarceo) [inline src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py:118]:
```suggestion
        self.vector_stores_config = vector_stores_config or VectorStoresConfig()
        self.file_processor_api = file_processor_api
```

File: src/llama_stack/providers/registry/file_processors.py (hunk 9 above)
Comment (franciscojavierarceo) [reply src/llama_stack/providers/registry/file_processors.py]: actually i think this is what gets autogenerated into the docs, right?

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        try:
+            from pypdf import PdfReader  # type: ignore[import-not-found]
+        except ImportError as err:
+            raise ImportError("PyPDF is required for PDF processing. Install with: pip install pypdf") from err
+
+        pdf_bytes = io.BytesIO(pdf_data)
+        reader = PdfReader(pdf_bytes)
+
+        # Handle password-protected PDFs
+        if reader.is_encrypted:
```
Comment (franciscojavierarceo) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: We shouldn't accept password protected files

File: tests/integration/responses/recordings/175a594469fd2c66751bd037a78fa3d6453e7b8bbb2a55f7baf6147997026931.json
Comment (alinaryan) [reply tests/integration/responses/recordings/175a594469fd2c66751bd037a78fa3d6453e7b8bbb2a55f7baf6147997026931.json:1]: looks like this one was related to test_response_file_search_filter_by_category, not my work so I'll remove this one

Comment (alinaryan) [reply tests/integration/responses/recordings/175a594469fd2c66751bd037a78fa3d6453e7b8bbb2a55f7baf6147997026931.json:1]: my b, actually this tests the file search filtering functionality that goes through the vector store code I modified so it's relevant and should stay

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        if file:
+            # Read from uploaded file
+            content = await file.read()
+            filename = file.filename or f"{uuid.uuid4()}.pdf"
+        elif file_id:
+            # Get file from file storage using Files API
+            if not self.files_api:
+                raise ValueError("Files API not available - cannot process file_id")
+
+            from llama_stack_api.files import RetrieveFileContentRequest, RetrieveFileRequest
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: keep imports at the top

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+            file_info = await self.files_api.openai_retrieve_file(RetrieveFileRequest(file_id=file_id))
+            filename = file_info.filename
+
+            # Get file content
+            content_response = await self.files_api.openai_retrieve_file_content(
+                RetrieveFileContentRequest(file_id=file_id)
+            )
+            content = content_response.body
+        else:
+            raise ValueError("Neither file nor file_id provided")
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: there's already a test for this at the top of the method

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+            raise ValueError("Either file or file_id must be provided")
+        if file and file_id:
+            raise ValueError("Cannot provide both file and file_id")
+
+        start_time = time.time()
+
+        # Get PDF content
+        if file:
+            # Read from uploaded file
+            content = await file.read()
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:60]: we should have a size limit

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        if self.config.clean_text:
+            text_content = self._clean_text(text_content)
+
+        # Extract metadata if configured
+        pdf_metadata = {}
+        if self.config.extract_metadata:
+            pdf_metadata = self._extract_pdf_metadata(content)
+
+        # Create document ID - prefer file_id for stability, fallback to filename
+        document_id = file_id if file_id else (filename or str(uuid.uuid4()))
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: please avoid using user input as an id

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+        # Create chunks for non-empty text
+        chunks = self._create_chunks(text_content, document_id, chunking_strategy, document_metadata)
+
+        return ProcessFileResponse(chunks=chunks, metadata=response_metadata)
+
+    def _extract_pdf_text(self, pdf_data: bytes) -> str:
+        """Extract text from PDF using PyPDF."""
+        try:
+            from pypdf import PdfReader  # type: ignore[import-not-found]
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: keep imports at the top

Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: do not type ignore here

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+        # Extract text from all pages
+        text_parts = []
+        for page_num, page in enumerate(reader.pages):
+            try:
+                page_text = page.extract_text()
+                if page_text and page_text.strip():  # Only add non-empty pages
+                    text_parts.append(page_text)
+            except Exception as e:
+                # Log warning but continue processing other pages
+                log.warning(f"Failed to extract text from page {page_num + 1}: {e}")
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: this silently ignores parts of the input. what should the user do here?

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+                continue
+
+        return "\n".join(text_parts)
+
+    def _extract_pdf_metadata(self, pdf_data: bytes) -> dict[str, Any]:
+        """Extract metadata from PDF."""
+        try:
+            from pypdf import PdfReader  # type: ignore[import-not-found]
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: imports at the top

Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: do not type ignore

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+        text_content = self._extract_pdf_text(content)
+
+        # Clean text if configured
+        if self.config.clean_text:
+            text_content = self._clean_text(text_content)
+
+        # Extract metadata if configured
+        pdf_metadata = {}
+        if self.config.extract_metadata:
+            pdf_metadata = self._extract_pdf_metadata(content)
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: don't double parse the file

File: src/llama_stack/providers/registry/file_processors.py [hunk 20]
```diff
@@ -4,8 +4,22 @@
...
 def available_providers() -> list[ProviderSpec]:
-    return []
+    return [
+        InlineProviderSpec(
+            api=Api.file_processors,
+            provider_type="inline::pypdf",
+            pip_packages=["pypdf"],
+            module="llama_stack.providers.inline.file_processor.pypdf",
+            config_class="llama_stack.providers.inline.file_processor.pypdf.PyPDFFileProcessorConfig",
+            optional_api_dependencies=[Api.files],
```
Comment (mattf) [inline src/llama_stack/providers/registry/file_processors.py]: let's make files a required api and if someone wants to only support part of the file-processing api later we can discuss making it optional

File: src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py [hunk 21]
```diff
@@ -940,34 +943,66 @@ async def openai_attach_file_to_vector_store(
             max_chunk_size_tokens = chunking_strategy.static.max_chunk_size_tokens
             chunk_overlap_tokens = chunking_strategy.static.chunk_overlap_tokens
         else:
-            # Default values from OpenAI API spec
-            max_chunk_size_tokens = 800
-            chunk_overlap_tokens = 400
+            # Use vector-io defaults for auto chunking
+            from llama_stack_api.vector_io import VectorStoreChunkingStrategyStaticConfig
```
Comment (mattf) [inline src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py]: imports at top

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 14 above)
Comment (alinaryan) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:60]: I'll add one here. Noticed the files api doesn't have a limit either, maybe we should add one there too

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 17 above)
Comment (alinaryan) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py]: changing _extract_pdf_text to return both the text and a list of failed pages in the metadata. They'll still get chunks from the pages that succeeded.

File: src/llama_stack/providers/registry/file_processors.py (hunk 20 above)
Comment (alinaryan) [reply src/llama_stack/providers/registry/file_processors.py]: I designed the file processors api so that the user can either direct upload to the file processors api or upload a file via the files api. If files becomes a required api here then a user can't deploy the file processors provider without also having a files api configured.

Comment (alinaryan) [reply src/llama_stack/providers/registry/file_processors.py]: if we don’t see a strong use case for direct uploads to the processor itself, I’m fine simplifying the design and requiring files, or restructuring so uploads always go through the Files API.

Comment (cdoern) [reply src/llama_stack/providers/registry/file_processors.py]: the files API is enabled in most baked-in distros, right? so this shouldn't be a huge deal to make files a required API here.

Comment (mattf) [reply src/llama_stack/providers/registry/file_processors.py]: > I designed the file processors api so that the user can either direct upload to the file processors api or upload a file via the files api. If files becomes a required api here then a user can't deploy the file processors provider without also having a files api configured.

imho those are two different types of users. the first is an api user, the second is a stack admin.

if we make file_id optional for the stack admin, the api user doesn't get a consistent api across stack deployments.

improving the api user's experience at the cost of the stack admin's experience is a worthwhile endeavor, just like making stack admin's experience better at the cost of the stack developer's (you, charlie, me, etc).

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 14 above)
Comment (franciscojavierarceo) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:60]: yeah we should, we can do that as a follow up for sure

Comment (mattf) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:60]: also, the files api shouldn't read the entire file into memory

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+        if not file and not file_id:
+            raise ValueError("Either file or file_id must be provided")
+        if file and file_id:
+            raise ValueError("Cannot provide both file and file_id")
+
+        start_time = time.time()
+
+        # Get PDF content
+        if file:
+            # Read from uploaded file
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:59]:
```suggestion
            # Read from uploaded file (TODO: read in chunks to avoid reading more than max_file_size_bytes)
```

File: src/llama_stack/providers/inline/file_processor/pypdf/adapter.py [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+from llama_stack_api.vector_io import VectorStoreChunkingStrategy
+
+from .config import PyPDFFileProcessorConfig
+from .pypdf import PyPDFFileProcessor
+
+
+class PyPDFFileProcessorAdapter:
+    """Adapter for PyPDF file processor."""
+
+    def __init__(self, config: PyPDFFileProcessorConfig, files_api=None) -> None:
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/adapter.py:21]:
```suggestion
    def __init__(self, config: PyPDFFileProcessorConfig, files_api) -> None:
```

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+
+# Window size for single-chunk mode (no chunking strategy)
+# Large enough to fit any reasonable document in one chunk
+SINGLE_CHUNK_WINDOW_TOKENS = 1_000_000
+
+
+class PyPDFFileProcessor:
+    """PyPDF-based file processor for PDF documents."""
+
+    def __init__(self, config: PyPDFFileProcessorConfig, files_api=None) -> None:
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:36]:
```suggestion
    def __init__(self, config: PyPDFFileProcessorConfig, files_api) -> None:
```

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 25]
```diff
@@ -0,0 +1,227 @@
...
+            content = await file.read()
+            if len(content) > self.config.max_file_size_bytes:
+                raise ValueError(
+                    f"File size {len(content)} bytes exceeds maximum of {self.config.max_file_size_bytes} bytes"
+                )
+            filename = file.filename or f"{uuid.uuid4()}.pdf"
+        elif file_id:
+            # Get file from file storage using Files API
+            if not self.files_api:
+                raise ValueError("Files API not available - cannot process file_id")
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:69]:
```suggestion

```

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+
+            # Get file metadata
+            file_info = await self.files_api.openai_retrieve_file(RetrieveFileRequest(file_id=file_id))
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:72]: (make sure there is a test for file_id doesn't exist and user gets a ValueError in response)

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py [hunk 27]
```diff
@@ -0,0 +1,227 @@
...
+        # Extract text from all pages
+        text_parts = []
+        failed_pages = []
+        for page_num, page in enumerate(reader.pages):
+            try:
+                page_text = page.extract_text()
+            except Exception as e:
+                failed_pages.append(f"page {page_num + 1}: {e}")
+                continue
+            if page_text and page_text.strip():
```
Comment (mattf) [inline src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:143]: nit: checking page_text.strip() feels like something to do only if cleaning is configured

File: src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py [hunk 28]
```diff
@@ -954,28 +958,57 @@ async def openai_attach_file_to_vector_store(
...
+                    # Fall back to legacy chunking path
+                    chunks = await self._legacy_chunk_file(
+                        file_id, file_response, max_chunk_size_tokens, chunk_overlap_tokens, chunk_attributes
+                    )
+            else:
+                logger.debug(f"FileProcessor API not available, using legacy chunking for file {file_id}")
+                # Legacy chunking path when FileProcessor not available
+                chunks = await self._legacy_chunk_file(
+                    file_id, file_response, max_chunk_size_tokens, chunk_overlap_tokens, chunk_attributes
+                )
```
Comment (mattf) [inline src/llama_stack/providers/utils/memory/openai_vector_store_mixin.py:999]: imho, the file processor api should be a non-optional dependency

File: src/llama_stack/providers/utils/memory/vector_store.py [hunk 29]
```diff
@@ -109,6 +113,182 @@ def content_from_data_and_mime_type(data: bytes | str, mime_type: str | None, en
...
+            if doc.mime_type == "application/pdf":
+                return parse_pdf(r.content)
+            return str(r.text)
+        return doc.content
+    else:
+        # will raise ValueError if the content is not List[InterleavedContent] or InterleavedContent
+        return interleaved_content_as_str(doc.content)
+
+
+async def content_from_doc_with_processor(doc: RAGDocument, file_processor_api) -> str:
```
Comment (mattf) [inline src/llama_stack/providers/utils/memory/vector_store.py:226]: confirm this is used

File: src/llama_stack/providers/utils/memory/vector_store.py [hunk 30]
```diff
@@ -109,6 +113,182 @@ def content_from_data_and_mime_type(data: bytes | str, mime_type: str | None, en
...
+                    r.content, doc.mime_type, file_processor_api, filename=doc.content.split("/")[-1]
+                )
+            return str(r.text)
+        return doc.content
+    else:
+        # will raise ValueError if the content is not List[InterleavedContent] or InterleavedContent
+        return interleaved_content_as_str(doc.content)
+
+
+async def content_from_data_with_processor(data_url: str, file_processor_api) -> str:
```
Comment (mattf) [inline src/llama_stack/providers/utils/memory/vector_store.py:261]: confirm this is used

**Review by mattf — APPROVED** [review]
looking good

File: src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py (hunk 27 above)
Comment (alinaryan) [reply src/llama_stack/providers/inline/file_processor/pypdf/pypdf.py:143]: I added it to filter out pdf pages with only whitespace, but not a huge deal

**Comment by cdoern** [comment]
a bunch of these failures are from pre-commit https://github.com/llamastack/llama-stack/actions/runs/22112540814/job/63912222690?pr=4743 the backwards compatibility is failing because it detects a diff in uv.lock from pre-commit I think. Are you running pre-commit locally?

the integration failures are from improper rebases:
```
>>>>>>> ddfa12cb5 (feat: enable inline pypdf file processor provider\)
```

still in the code ^

**Comment by cdoern** [comment]
@alinaryan what is the status on this, is this still ready for review? I see it has an approval in the past week but has anything functional changed since then?

**Comment by cdoern** [comment]
remaining failure is a registry access issue, nothing with this PR

**Comment by alinaryan** [comment]
Thanks for the patience here. I ended up dropping the changes made post-approval to get this PR back to a stable state and made sure to add relevant recordings. I’ll follow up with updates in a separate PR.

Apologies for the churn!

**Review by franciscojavierarceo — APPROVED** [review]
great work here @alinaryan thank you!

File: src/llama_stack/providers/registry/file_processors.py (hunk 9 above)
Comment (alinaryan) [reply src/llama_stack/providers/registry/file_processors.py]: shortened it
